### DISCUSSION TITLE
Are the "efficient-kan" and "official-kan" equivalent in terms of algorithms?

### DISCUSSION THREAD

**Discussion by yuedajiong** [Blealtan/efficient-kan#35] [General]
as title

**Comment by Indoxer** [comment]
As I know almost the same, only official version looks to have additional bias after each layer. Also, I am not sure if initialization is the same. + regularization loss is changed because of optimizations.

**Comment by yuedajiong** [comment]
@Indoxer   Thanks, you are so kindly.

**Comment by WhatMelonGua** [comment]
No, I'm not quite sure
I tried the official tutorial on the following link: [Tutorial](https://github.com/KindXiaoming/pykan/blob/master/tutorials/Example_7_continual_learning.ipynb)

*Including the use of the official LBFGS training strategy
The results showed that after completing all the one-time training, the model was almost identical to the official one
But if training is conducted in phases, it cannot be perfectly fitted(But the model is still effective, just slightly underperforming)
![image](https://github.com/Blealtan/efficient-kan/assets/92420346/53262b98-af5d-4891-a600-7059d292cd5e)
official KAN
![image](https://github.com/Blealtan/efficient-kan/assets/92420346/7f03f983-3f21-4b9c-be7c-acc0b23a7ab2)
Eff-KAN

**Comment by WhatMelonGua** [comment]
I think this is acceptable, after all, the model is very efficient, and some losses are normal. It's strange if there are no losses at all. While it effectively retains the characteristics of the official model, it also combines training optimization

**Comment by Indoxer** [comment]
@WhatMelonGua, are you sure, that you didn't train spline_scaler and base_weights? Also did you have the same parameters in LBFGS optimizer (number of steps, etc.)?

**Comment by Indoxer** [comment]
(spline_scaler not trained, base_weights not trained)
![spline_scaler not trained, base_weights not trained](https://github.com/Blealtan/efficient-kan/assets/77686332/9c2abc73-f4ca-40e0-915e-722ed926b080)
(spline_scaler trained, base_weights trained):
![spline_scaler trained, base_weights trained](https://github.com/Blealtan/efficient-kan/assets/77686332/76b70c2c-22f5-4bde-880c-52b69e9dd7cc)

(I am using my modified version (but the same algorithm as efficient kan), so I am not sure)

**Comment by WhatMelonGua** [comment]
> @WhatMelonGua, are you sure, that you didn't train spline_scaler and base_weights? Also did you have the same parameters in LBFGS optimizer (number of steps, etc.)?

Oh, yes, forgive me for forgetting
There are no such parameters, so for that reg_ variable (I don't know what it is), I simply took the default value of 1 and fixed many errors (perhaps I was fixing it blindly, just making it work)
And then the result was that the official "LBFGS" cannot be directly migrated here

**Comment by WhatMelonGua** [comment]
> (spline_scaler not trained, base_weights not trained) ![spline_scaler not trained, base_weights not trained](https://private-user-images.githubusercontent.com/77686332/330014808-9c2abc73-f4ca-40e0-915e-722ed926b080.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU2MTQ4NzgsIm5iZiI6MTcxNTYxNDU3OCwicGF0aCI6Ii83NzY4NjMzMi8zMzAwMTQ4MDgtOWMyYWJjNzMtZjRjYS00MGUwLTkxNWUtNzIyZWQ5MjZiMDgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTEzVDE1MzYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjNTFmZGQ2Yzk5ODJlOTlmNjc3NTk0MGYyMWE2MTA2NDkwY2IxOTQ5ZTBjYTkyMDgyZmNmYWNiM2ZmZTc5YjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LdqnOPNje_Od8O9RA5Yn6n-F8MsZ3ND2OS75H94umqs) (spline_scaler trained, base_weights trained): ![spline_scaler trained, base_weights trained](https://private-user-images.githubusercontent.com/77686332/330017625-76b70c2c-22f5-4bde-880c-52b69e9dd7cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU2MTQ4NzgsIm5iZiI6MTcxNTYxNDU3OCwicGF0aCI6Ii83NzY4NjMzMi8zMzAwMTc2MjUtNzZiNzBjMmMtMjJmNS00YmRlLTg4MGMtNTJiNjllOWRkN2NjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTEzVDE1MzYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0MTI0N2Q5YWJhOTNhN2M1ODM4OTUzOTdjZDI4MWJjMjQxZTM5MTJlMmY5MGNmMzVlZWZlY2UxMmExYjdjOWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7OAk-Rmfami2ZvF8ADuBHWYxiJVch7I_kq2qGatr0a4)
> 
> (I am using my modified version (but the same algorithm as efficient kan), so I am not sure)

This may seem like our operations are similar
What a coincidence! 🤗

**Comment by Indoxer** [comment]
> > @WhatMelonGua, are you sure, that you didn't train spline_scaler and base_weights? Also did you have the same parameters in LBFGS optimizer (number of steps, etc.)?
> 
> Oh, yes, forgive me for forgetting There are no such parameters, so for that reg_ variable (I don't know what it is), I simply took the default value of 1 and fixed many errors (perhaps I was fixing it blindly, just making it work) And then the result was that the official "LBFGS" cannot be directly migrated here

reg_ is regularization loss. `loss = train_loss + lamb * reg_` for continual learning `lamb=0.0` so `loss = train_loss`

**Reply by WhatMelonGua** [reply]
Thank you, let me know those knowledge!

**Comment by Indoxer** [comment]
Here are [my results and code](https://github.com/Indoxer/LKAN/blob/main/examples/continual_learning_lbfgs.ipynb), so you can compare

**Comment by Blealtan** [comment]
AFAIK the only difference is that the "efficient" regularization loss is different from the official one. But I'm not sure if the parallel associativity will introduce numerical error that's large enough to break some important features.

**Comment by Blealtan** [comment]
Just found that I missed the bias term after each layer. Will update that soon.

I scanned over this long thread few days ago and totally missed the comment by @Indoxer lol

**Comment by mtom365** [comment]
Hi is there any plan to support update_grid_from_samples and initialize_from_another_model from the original KANs.  A lot of use cases work much better when we use these  APIs so I think its critical for KANs

**Comment by EladWarshawsky** [comment]
Its seems like the algos are somewhat equivalent. As someone who is much more interested in the symbolic representation aspect, is there enough consistency between the original KAN implementation and what's here to do formula approximation?
